### ISSUE TITLE
OnEntry not been called while loading statemachine with specific state

### ISSUE THREAD

**Issue by dporikos** [dotnet-state-machine/stateless#343]
Hi,
I have a daemon which runs periodically. The daemon gets tasks from DB and loads a state machine for each of them. Then stores back the state of each task in DB. 

When a transition from one state to another is kicked, OnEntry methods are called.
However, if a state is loaded from DB the OnEntry method is not called. Am I missing anything ?

Following code is placed on he constructor of my class:
`public TaskItem(int TaskID, State taskState)
        {
            _taskID = TaskID;
            _stateMachine = new StateMachine<State, Trigger>(taskState);
            _automateTrigger = _stateMachine.SetTriggerParameters<int>(Trigger.Automate);

            //Configure Open State
            _stateMachine.Configure(State.Open)
                .Permit(Trigger.Automate, State.InAutomationProgress)
                .Permit(Trigger.Assign, State.InManualProgress);

            //Configure InAutomationProgress State
            _stateMachine.Configure(State.InAutomationProgress)
                .Permit(Trigger.Fail, State.Failed)
                .Permit(Trigger.Complete, State.Completed)
                //.PermitReentry(Trigger.Automate)
                .OnEntryFrom(_automateTrigger, _taskID => OnAutomationProgress(_taskID));

            //Configure InManualProgress State
            _stateMachine.Configure(State.InManualProgress)
                .Permit(Trigger.Complete, State.Completed)
                .Permit(Trigger.Cancel, State.Deferred)
                .OnEntry(t => CheckTicketStatus());
        }`

The CheckTicketStatus() and OnAutomationProgress(taskID) methods are not called.

A quick workaround is to PermitReentry on every state and call the Fire event again inside my constructor but I am no sure if it the right approach:

`switch (taskState)
            {
                case State.InAutomationProgress:
                    _stateMachine.Fire(_automateTrigger, _taskID);
                    break;
                case State.InManualProgress:
                    _stateMachine.Fire(Trigger.Assign);
                    break;
                default:
                    break;
            }`

Any ideas ?

Thank you.
Dimitris

**Comment by uygary** [comment]
I don't know whether this was the intended design or not, but it sounds about right to me.
The state machine is not entering a new state here. It's being instantiated with that state.
The entry logic did already run when the state machine was initially transitioned to that particular state, before the state was persisted to db. If you intend to run it again, that just doesn't sound right to me.
However I have no idea about your actual business logic and requirements, so take any comment I have with a backpack full of salt. If that's indeed what you need, I suggest you have an initial state like RetrievingPersistedState, and then actually transition to the persisted states after you retrieve them from db.

**Comment by vlm---** [comment]
Try to use OnActivate/OnActivateAsync, thats exactly what they are for.

**Comment by dporikos** [comment]
I also tried the OnActivate method but not triggers though.

As uygary mentioned it seems to be right not to trigger the event again since it was already triggered during the state transition.
Since the program is going to run on schedule basis, I need to determine changes applied in external systems in order to determine possible transitions to other states. The option to use a RetrievingPersistedState sounds quite reasonable. Therefore, I managed to resolve my issue by triggering again the triggers in my constructor.

Thank you both for your help !
D.
